go/vt/vtgate/planbuilder/operators/filter.go

```diff
@@ -60,8 +67,15 @@ func (f *Filter) UnsolvedPredicates(st *semantics.SemTable) []sqlparser.Expr {
 	return result
 }
 
-func newFilter(op Operator, expr ...sqlparser.Expr) Operator {
-	return &Filter{
-		Source: op, Predicates: expr,
+func (f *Filter) AddPredicate(ctx *plancontext.PlanningContext, expr sqlparser.Expr) (Operator, error) {
+	newSrc, err := f.Source.AddPredicate(ctx, expr)
+	if err != nil {
+		return nil, err
 	}
+	f.Source = newSrc
+	return f, nil
```

Comment:
I think here we can check if the source ends up creating a filter for this predicate, we can coalesce the two filters into one.

Reply:
This happens in filter.compact, so all good.